docs(ReadMe): Updated with correct addresses and user guide on new implementation (#622)
WavesHQ/quantum: README.md

TODO

> ####

Reviewer: ```suggestion
## Contract Operations
```

To change the state of any smart contract, users need to approve the smart contract of the respective token via the `approve()` function first. Once approved, user will be able to bridge the token over to DefiChain.

We are implementing a TimeLock contract that will work as an Admin address for ADMIN ONLY tx. There will be 3 days delay on every operational tx except when calling `Withdraw()` function, TimeLock contract is not able to call this function.

> ### TimeLock Contract Account Permission

Reviewer 2: ## Timelock Contract Operation

> Gnosis safe will be implemented with both proposers and executors roles. Gnosis Safe has the authority to grant roles to addresses, as long as it holds the corresponding role itself. When revoking privileges, either the revokeRole() or renounceRole() functions must be used.

Reviewer 2: ```suggestion
Gnosis safe will be implemented with both proposers and executors roles. Only the Timelock smart contract has the role (TIMELOCK_ADMIN_ROLE) to grant roles for now. If we want to grant new roles to new addresses, have to go through a round of scheduling and executing the grantRole functions through the Timelock contract. When revoking privileges, either the revokeRole() or renounceRole() functions must be used.
```

Salt can be `0x0000000000000000000000000000000000000000000000000000000000000000` for 1st transaction,
`0x0000000000000000000000000000000000000000000000000000000000000001` for 2nd transaction and so on.

The reason behind choosing different salt is to avoid having the same transaction again.

After scheduling the transaction using a timelock, the developer must call the execute() function with the provided arguments(same as above). If the execute() function is called before the specified `delay` time has passed, the transaction will revert with the error message "TimelockController: operation is not ready". This is because the timelock is designed to ensure that the specified delay time has elapsed before the transaction can be executed. Once the delay time has passed, the transaction can be executed normally.

> ## Operational Transactions

Reviewer 2: ```suggestion
## Bridge Contract operation
```

### Bridge Contract Account Permission

There are only two roles: DEFAULT_ADMIN_ROLE and WITHDRAW_ROLE.

> The TimeLock contract is assigned the DEFAULT_ADMIN_ROLE, while another Gnosis Safe is assigned the WITHDRAW_ROLE. The DEFAULT_ADMIN_ROLE has the ability to grant both roles to other addresses, but this change will only take effect after three days.

Reviewer 2: actually, from the perspective of the bridge contract, the change will happen instantly after default_admin_role call the grantRole function?

To change the state of any smart contract, users need to approve the smart contract of the respective token via the `approve()` function first. Once approved, user will be able to bridge the token over to DefiChain.

> We are implementing a TimeLock contract that will work as an Admin address for ADMIN ONLY tx. There will be 3 days delay on every operational tx except when calling `Withdraw()` function.

Reviewer 2: Timelock cannot withdraw iirc?

Author: Yes, this is what it says as well.

Reviewer 2: hmm, the way it writes here indicates that Timelock contract can withdraw?

### TimeLock Contract Account Permission

> Gnosis safe will be implemented with both proposers and executors roles. Gnosis Safe has the authority to grant roles to addresses, as long as it holds the corresponding role itself. When revoking privileges, either the revokeRole() or renounceRole() functions must be used.

Reviewer 2: What do you mean by the corresponding role itself?

Author: If multi-sig wallet wants to give other wallet a proposers role, they can only do so if the executing account hold that role as well.

To execute only Admin transactions, the developer will need to follow these steps:

- First, create a transaction using Safe [guide](<(https://help.safe.global/en/articles/3738081-contract-interactions),>).

> - After providing the contract address and ABI, the developer can select the contract method (in this case, we will try to change the transaction fee).

Reviewer 2: should we say clearly we should provide the TimelockController address and its abi here?

Author: Guide is pretty explanatory. They will have to look this as a guide anyway.

Salt can be `0x0000000000000000000000000000000000000000000000000000000000000000` for 1st transaction,
`0x0000000000000000000000000000000000000000000000000000000000000001` for 2nd transaction and so on.

The reason behind choosing different salt is to avoid having the same transaction again.

After scheduling the transaction using a timelock, the developer must call the execute() function with the provided arguments(same as above). If the execute() function is called before the specified `delay` time has passed, the transaction will revert with the error message "TimelockController: operation is not ready". This is because the timelock is designed to ensure that the specified delay time has elapsed before the transaction can be executed. Once the delay time has passed, the transaction can be executed normally.

### Bridge Contract Account Permission

There are only two roles: DEFAULT_ADMIN_ROLE and WITHDRAW_ROLE.

> The TimeLock contract is assigned the DEFAULT_ADMIN_ROLE, while another Gnosis Safe is assigned the WITHDRAW_ROLE. The DEFAULT_ADMIN_ROLE has the ability to change both roles to other addresses, but this change will only take effect after three days.

Reviewer 2: ```suggestion
The TimeLock contract is assigned the DEFAULT_ADMIN_ROLE, while another Gnosis Safe is assigned the WITHDRAW_ROLE. The DEFAULT_ADMIN_ROLE has the ability to grant both roles to other addresses, but this change will only take effect after three days.
```

Finally, both addresses can renounce their own roles by calling the renounceRole() function.

### Bridge ERC20 tokens - to transfer ERC20 tokens from an EOA to the Bridge

Once approved, user will call the `bridgeToDeFiChain()` function with following arguments: `_defiAddress`- address on Defi Chain that receiving funds, `_tokenAddress` - ERC20 token's address and `_amount` amount to bridge over to Defi chain.

### Add supported token

> Only address with the Admin role can call the `addSupportedTokens()` function. This sets the `_tokenCap` for an ERC20 token and ETH identified by its `_tokenAddress`. All added tokens will be instantly supported by the bridge.

Reviewer 2: ```suggestion
Only address with the Admin role of the BridgeProxy contract can call the `addSupportedTokens()` function. This sets the `_tokenCap` for an ERC20 token and ETH identified by its `_tokenAddress`. All added tokens will be instantly supported by the bridge.
```

Author: I don't think we need to add of `of the BridgeProxy contract` as the function we are calling is of BridgeProxy contract.

In case of ETH, address(0) will be used as an address.

`_tokenCap` represent the maximum balance of tokens the contract can hold per `_tokenAddress`

### Remove supported token

> Only address with the Admin role can call the `removeSupportedTokens()` function. This also sets the `_tokenCap` to `0`.

Reviewer 2: ```suggestion
Only address with the Admin role of BridgeProxy contract can call the `removeSupportedTokens()` function. This also sets the `_tokenCap` to `0`.
```

### Withdraw

> `withdraw()` function when called will withdraw an ERC20 token and ETH (address == 0x0). Only the address with the WITHDRAW role can call this function.

Reviewer 2: ```suggestion
`withdraw()` function when called will withdraw an ERC20 token and ETH (address == 0x0). Only the address with the WITHDRAW role of BridgeProxy contract can call this function.
```

### FlushFund

> `flushFund` function to flush the excess funds `(token.balanceOf(Bridge) - tokenCap)` across supported tokens to a hardcoded address (`flushReceiveAddress`) anyone can call this function. This applies to all tokens and ETH.

Reviewer 2: should change documentation about flushFund here?

### Change Flush Receive Address

> Admin address can change `flushReceiveAddress`.

Reviewer 2: Admin address of BridgeProxy contract

`changeFlushReceiveAddress` function will reset the `flushReceiveAddress` to new address.

### Change relayer address

> Admin address can change `relayerAddress`.

Reviewer 2: Admin address of BridgeProxy contract

The relayer address will primarily be used for verifying the signature that is signed by the server. The server will need to sign with the corresponding private key of the relayer address.

### Transaction fee change

> Only address with Admin role can change `transactionFee`.

Reviewer 2: Admin role of bridgeproxy contract

Initial fee will be set to 0%. This means that if the user bridges `X` tokens, 100% of X will be bridged to defiChain. If in future, fee > 0, respected amount will be sent to `communityWallet`.

### Change Tx Fee Address

> Only address with admin role can change `communityWallet`. This is where the tx fees upon bridging will go.

Reviewer 2: admin role of bridgeProxy contract

### Change Token Cap

> Only address with admin role can change `tokenCap`.

Reviewer 2: admin role of bridgeproxy contract

We are implementing a TimeLock contract that will work as an Admin address for ADMIN ONLY tx. There will be 3 days delay on every operational tx except when calling `Withdraw()` function, TimeLock contract is not able to call this function.

> ### TimeLock Contract Operations

Reviewer 2: should be two hash symbol?

We are implementing a TimeLock contract that will work as an Admin address for ADMIN ONLY tx. There will be 3 days delay on every operational tx except when calling `Withdraw()` function.

### TimeLock Contract Account Permission

Gnosis safe will be implemented with both proposers and executors roles. Gnosis Safe has the authority to grant roles to addresses, as long as it holds the corresponding role itself. When revoking privileges, either the revokeRole() or renounceRole() functions must be used.

To execute only Admin transactions, the developer will need to follow these steps:

- First, create a transaction using Safe [guide](<(https://help.safe.global/en/articles/3738081-contract-interactions),>).
- After providing the contract address and ABI, the developer can select the contract method (in this case, we will try to change the transaction fee).
- Select "Schedule". According to the Schedule() function, the following arguments need to be provided: address target, uint256 value, bytes calldata data, bytes32 predecessor, bytes32 salt, and uint256 delay.

> - The target address is the Proxy Bridge. The value will usually be 0. The data is the encoded data of the function and arguments (BridgeV1Interface.encodeFunctionData('changeTxFee', [100])). The predecessor is ZERO_BYTES32, and the salt will be in incremented order (e.g., 0x0...1, 0x0...2, 0x0...3, and so on).

Reviewer: Let's mention that predecessor should almost always be 0x0 unless we have a prerequisite transaction, and that the delay represents the minimum amount of time before the tx can be executed.

Is the delay in seconds or ms?

Author: Always in seconds. I will mentions this as well.

Salt can be `0x0000000000000000000000000000000000000000000000000000000000000000` for 1st transaction,
`0x0000000000000000000000000000000000000000000000000000000000000001` for 2nd transaction and so on.

> The reason behind choosing different salt is to avoid having he same operation identifier again.

Reviewer 2: "the same operation identifier"

After scheduling the transaction using a timelock, the developer must call the execute() function with the provided arguments(same as above). If the execute() function is called before the specified `delay` time has passed, the transaction will revert with the error message "TimelockController: operation is not ready". This is because the timelock is designed to ensure that the specified delay time has elapsed before the transaction can be executed. Once the delay time has passed, the transaction can be executed normally.

## Bridge Contract operations

### Bridge Contract Account Permission

There are only two roles: DEFAULT_ADMIN_ROLE and WITHDRAW_ROLE.

> The TimeLock contract is assigned the DEFAULT_ADMIN_ROLE, while another Gnosis Safe is assigned the WITHDRAW_ROLE. The DEFAULT_ADMIN_ROLE has the ability to grant both roles to other addresses, but this changes will happen instantly once executed.

Reviewer 2: "these changes will happen ..."

### Withdrawing funds

> ETH and ERC20 tokens can be withdrawn by the address with WITHDRAW_ROLE only via the `withdraw(address _tokenAddress, uint256 amount)` function.

Reviewer 2: WITHDRAW_ROLE of bridgeproxy contract

Author: `WITHDRAW_ROLE ` is only in BridgeProxy contract.

Reviewer 2: should we remove this one, cos it is a duplicate of a part before

## Admin address - Gnosis safe

> Admin address will be Gnosis safe, ideally will be with at least 3 owners with a 2/3 quorum.

Reviewer 2: Admin address of TimelockController contract

Reviewer 2: and how about withdraw_role of BridgeProxy? 
I think let this be discussed off-Github
